### DISCUSSION TITLE
How to prevent race condition when save data on Drift

### DISCUSSION THREAD

**Discussion by rizaldi-wiratama** [simolus3/drift#3254] [Q&A] [answered]
Hello everyone, firstly thank you so much for this awesome persistence database 

I have some question regarding how to prevent race condition when save data on Drift

There's case that i have `Order` class that already implements `Insertable<Order>` (custom row class) that consist of several field.
When saving an order using Manager API , first I check if field `dateLastActivity ` is already changed or not
if already changed then we throw an error "Order already changed"
but if not already changed then we update date last activity & save order with mode replace just like these code
```
  ///Save order by check first if local order already exist and date last activity is changed
  ///- if date last activity is changed then throw error
  ///- if not changed then update date last activity & save order with mode upsert
  saveOrder(Order updatedOrder) async {
    ///Get local order, can return null if order is not found with filter by [uuid]
    Order? localOrder = await database.managers.orders.filter((f) => f.uuid(updatedOrder.uuid)).getSingleOrNull();

    ///Check if date last activity is changed or not
    if(localOrder != null && localOrder.dateLastActivity != updatedOrder.dateLastActivity) throw("Order already changed");

    ///Update date last activity before save
    updatedOrder.dateLastActivity = dateFormat.format(DateTime.now());
    
    ///Save order with mode replace so it will behave like upsert (update/insert)
    await database.managers.orders.create((o) => order, mode: InsertMode.replace);
  }
```

This `saveOrder` function can be called on several service

My question, when there's case several service access `saveOrder` function with same order at the same time, is this may lead to race condition?
If yes, then how to ensure that this not lead to race condition?

I already research some way to prevent race condition such as Acquiring a Read Lock on SQLite in [here](https://www.sqlite.org/atomiccommit.html#_acquiring_a_read_lock) , is Drift also support like this?

and I already know there's `onConflict` callback inside `database.managers.orders.create` , but how to check if field `dateLastActivity` is changed or not inside that callback? , is there any way that we can check field  `dateLastActivity` & update that field before save order? or do we need Transaction for this case?

Thanks in advance

**Comment by simolus3** [comment, marked as answer]
Yes! You can run multiple statements atomically (with an implicit lock on the database) by using [transactions](https://drift.simonbinder.eu/docs/transactions/):

```dart
saveOrder(Order updatedOrder) async {
  await database.transaction(() async {
    Order? localOrder = await database.managers.orders....;
    // Rest of your update logic here
  })
}
```

Transactions essentially behave as if all their statements inside ran as one (which also means that no concurrent invocation of `saveOrder` can cause race conditions).

**Comment by rizaldi-wiratama** [comment]
Thank you so much @simolus3 for the answer , I will try to do it later using transactions 👍
